core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java

```diff
@@ -150,12 +157,49 @@ public void generateReportMoveThenRemove() {
     removeBlock(block1);
 
     // The block should not be in the added blocks list
-    BlockHeartbeatReport report = mReporter.generateReport();
+    BlockHeartbeatReport report = mReporter.generateReportAndClear();
     assertEquals(null, report.getAddedBlocks().get(MEM_LOC));
 
     // The block should be in the removed blocks list
     List<Long> removedBlocks = report.getRemovedBlocks();
     assertEquals(1, removedBlocks.size());
     assertTrue(removedBlocks.contains(block1));
   }
+
+  @Test
+  public void generateAndRevert() {
+    mReporter.onCommitBlockToMaster(1, MEM_LOC);
+    mReporter.onCommitBlockToMaster(2, MEM_LOC);
+    mReporter.onCommitBlockToMaster(3, SSD_LOC);
+    mReporter.onRemoveBlockByClient(4);
+    mReporter.onStorageLost(HDD_LOC);
+    BlockHeartbeatReport originalReport = mReporter.generateReportAndClear();
+    mReporter.revert(originalReport);
+    BlockHeartbeatReport newReport = mReporter.generateReportAndClear();
+    assertEquals(originalReport.getAddedBlocks(), newReport.getAddedBlocks());
+    assertEquals(originalReport.getRemovedBlocks(), newReport.getRemovedBlocks());
+    assertEquals(originalReport.getLostStorage(), newReport.getLostStorage());
+  }
+
+  @Test
+  public void generateUpdateThenRevert() {
+    mReporter.onCommitBlockToMaster(1, MEM_LOC);
+    mReporter.onCommitBlockToMaster(2, MEM_LOC);
+    mReporter.onCommitBlockToMaster(3, SSD_LOC);
+    mReporter.onRemoveBlockByClient(4);
+    BlockHeartbeatReport originalReport = mReporter.generateReportAndClear();
+    mReporter.onRemoveBlockByClient(1);
+    mReporter.onRemoveBlockByClient(3);
+    mReporter.onRemoveBlockByClient(5);
```

Comment:
need coverage on added blocks, because thats a map<location, list> so the logic may be more complicated

Reply:
yeah i updated the test